### ISSUE TITLE
[Bug]: panic: runtime error: index out of range [3442944] with length 1530368

### ISSUE THREAD

**Issue by larry88** [milvus-io/milvus#22796]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Environment

```markdown
- Milvus version: 2.2.3
- Deployment mode(standalone or cluster): cluster
- MQ type(rocksmq, pulsar or kafka):  pulsar
- SDK version(e.g. pymilvus v2.0.0rc2):
- OS(Ubuntu or CentOS): CentOS
- CPU/Memory: 96C/512G
- GPU: 
- Others:
```

### Current Behavior

when I insert data, I see the error message bellow:

[2023/03/16 09:25:55.370 +00:00] [INFO] [datanode/flow_graph_insert_buffer_node.go:451] ["insertBufferNode syncing BufferData"] [segmentID=439768816871499834] [flushed=false] [dropped=false] [auto=true] [position="channel_name:\"my-milvus-cluster-14-rootcoord-dml_12_439768816871299668v0\" msgID:\"\\010\\254\\232J\\020\\223/\\030\\000 \\000\" msgGroup:\"my-milvus-cluster-14-dataNode-93-my-milvus-cluster-14-rootcoord-dml_12_439768816871299668v0\" timestamp:440128684955533313 "] [channel=my-milvus-cluster-14-rootcoord-dml_12_439768816871299668v0]

 panic: runtime error: index out of range [3442944] with length 1530368

 

 goroutine 1834 [running]:

 github.com/milvus-io/milvus/internal/storage.(*DataSorter).Swap(0xc00034a0b0, 0x1183, 0x1759)

 	/go/src/github.com/milvus-io/milvus/internal/storage/data_sorter.go:95 +0x506

 sort.medianOfThree({0x41d6c10, 0xc00034a0b0}, 0xf9?, 0x0?, 0xc0005f4d40?)

 	/usr/local/go/src/sort/sort.go:94 +0x9e

 sort.doPivot({0x41d6c10, 0xc00034a0b0}, 0x0, 0x175a)

 	/usr/local/go/src/sort/sort.go:116 +0xd8

 sort.quickSort({0x41d6c10, 0xc00034a0b0}, 0x10?, 0x3b1f840?, 0xc001c64301?)

 	/usr/local/go/src/sort/sort.go:203 +0x6c

 sort.Sort({0x41d6c10, 0xc00034a0b0})

 	/usr/local/go/src/sort/sort.go:231 +0x53

 github.com/milvus-io/milvus/internal/storage.(*InsertCodec).Serialize(0xc00128b0e8, 0xc0005f52b0?, 0x66?, 0xc0009bc180)

 	/go/src/github.com/milvus-io/milvus/internal/storage/data_codec.go:286 +0x188

 github.com/milvus-io/milvus/internal/datanode.(*rendezvousFlushManager).flushBufferData(0xc000a378c0, 0xc001d1f1c0, 0x61a5f68b0eb683a, 0x0?, 0xc0?, 0xc0005f56d0?)

 	/go/src/github.com/milvus-io/milvus/internal/datanode/flush_manager.go:367 +0x2dc

 github.com/milvus-io/milvus/internal/datanode.(*insertBufferNode).Sync.func1()

 	/go/src/github.com/milvus-io/milvus/internal/datanode/flow_graph_insert_buffer_node.go:463 +0x63

 github.com/milvus-io/milvus/internal/util/retry.Do({0x41dd270, 0xc0019feac0}, 0xc0005f58c0, {0xc0005f5800, 0x1, 0x29230cc?})

 	/go/src/github.com/milvus-io/milvus/internal/util/retry/retry.go:37 +0x103

 github.com/milvus-io/milvus/internal/datanode.(*insertBufferNode).Sync(0xc000eae000, 0xc0005551f0?, {0xc0005551f0?, 0xc001c64300?, 0xc001c642a0?}, 0xc001c64300)

 	/go/src/github.com/milvus-io/milvus/internal/datanode/flow_graph_insert_buffer_node.go:462 +0x729

 github.com/milvus-io/milvus/internal/datanode.(*insertBufferNode).Operate(0xc000eae000, {0xc001990730?, 0x1?, 0x3e87978?})

 	/go/src/github.com/milvus-io/milvus/internal/datanode/flow_graph_insert_buffer_node.go:203 +0x985

 github.com/milvus-io/milvus/internal/util/flowgraph.(*nodeCtx).work(0xc001ad81e0)

 	/go/src/github.com/milvus-io/milvus/internal/util/flowgraph/node.go:127 +0x2ac

 created by github.com/milvus-io/milvus/internal/util/flowgraph.(*nodeCtx).Start

 	/go/src/github.com/milvus-io/milvus/internal/util/flowgraph/node.go:71 +0x79

### Milvus Log

[indexnode.log](https://github.com/milvus-io/milvus/files/10989292/indexnode.log)
[proxy.log](https://github.com/milvus-io/milvus/files/10989294/proxy.log)
[querycoord.log](https://github.com/milvus-io/milvus/files/10989296/querycoord.log)
[datacoord.log](https://github.com/milvus-io/milvus/files/10989297/datacoord.log)
[rootcoord.log](https://github.com/milvus-io/milvus/files/10989298/rootcoord.log)
[indexcoord.log](https://github.com/milvus-io/milvus/files/10989299/indexcoord.log)
[querynode.log](https://github.com/milvus-io/milvus/files/10989300/querynode.log)
[datanode.log](https://github.com/milvus-io/milvus/files/10989301/datanode.log)

**Comment by Cheung7788** [comment]
<img width="383" alt="image" src="https://user-images.githubusercontent.com/28556111/225582434-2db5cc66-3104-444c-bfc9-423254cc5b53.png">

**Comment by xiaofan-luan** [comment]
/assign @XuanYang-cn 
pls take a look into it

**Comment by larry88** [comment]
curl -X 'POST' 'http://172.19.80.95:31501/api/v1/entities'  -H 'accept: application/json' -H 'Content-Type: application/json' \
-d '{
"collection_name": "multimodal_txt",
"fields_data": [{ "field_name":"item_code","type":5,"field":[84949708] },{ "field_name":"category","type":5,"field":[109004002001] },{ "field_name":"vector","type":101,"field":[[0.1415857891624166, 0.06664296627817194, 0.07231860351614813, -0.05199616050274954, -0.024655456388627716, 0.04336064793099008, -0.0386309502326766, -0.1376799742889706, 0.08470736006785957, -0.13377415941552465, -0.004474141451711063, -0.09489909575325767, -0.06011293203662946, 0.012876983410892199, 0.03274171374380884, -0.045099956116821494, -0.010092564604627001, -0.01591314419142247, 0.06945027071846124, 0.07915377954467859, 0.0873315794359561, -0.10167324342439053, -0.028057787313543546, -0.08196108398496788, -0.03924123380665253, 0.056817400737159414, -0.06377463348048505, -0.013571180976289824, -0.029659781695230373, -0.016584456122795996, 0.08806391972472721, 0.09526526589764323, -0.08171697055537751, 0.09813359869533012, 0.025677681375037405, -0.05797693952771369, -0.04363527553927925, -0.07659058853397967, -0.07109803636819625, -0.04488635686592991, 0.047388519519231245, -0.004077457128626706, -0.0900778555188478, 0.04754109041272522, 0.024914826907567487, -0.026043851519422966, -0.0017946151347229796, -0.05498655001523162, 0.1194935237844878, 0.05568837612530394, -0.10972898660087285, -0.00633932062467501, 0.024670713477977113, 0.034664107001833026, 0.07067083786641311, -0.06499520062843693, -0.04326910539489369, -0.03734935472732714, -7.337706408976264E-4, 0.04180442481735145, 0.049036285168966263, -0.0024296914788916857, -0.06054013053841261, 0.03170423166804975, -0.09886593898410125, -0.082693424273739, 0.029614010427182178, -0.0021932065939760112, 0.026165908234218153, 0.0065948768712774325, -0.08922345851528149, -0.002509791197976027, -0.022107522467278193, -0.07152523486997941, 0.03182628838284494, -9.974322162169165E-4, -0.0033146026611567893, -0.07225757515875053, -0.00670549076906057, -0.03960740395103809, 0.051202791856580825, 0.016035200906217656, 0.08604998393060663, 0.025555624660242218, 0.0854397003566307, -0.009772165728289636, 0.09312927338872747, 0.022443178432964957, 0.03530490475450776, -0.010168850051373994, -0.1257184162390423, 0.03133806152366419, 0.10399232100549909, -0.04006511663152004, 0.07335608559190722, 0.11613696412762016, 0.04220110914043581, 0.07366122737889519, -0.05275901497021946, 0.010779133625349926, 0.034847192074025805, 0.04262830764221896, -0.05986881860703908, 0.0743325393102687, 0.03142960405976058, -0.035152333861013775, 0.01792707998554305, 0.008688912384482354, -0.04250625092742377, -0.008688912384482354, 0.010039164791904108, 0.04015665916761643, 0.10057473299123386, -0.09520423754024564, 0.0716472915847746, 0.0616081267928705, 0.05986881860703908, -0.05956367682005112, 0.0800081765482449, 0.04799880309320718, 0.011770844433060819, -0.05108073514178564, -6.293549356626815E-4, 0.012281956926265663, -0.04647309415826734, 0.03536593311190535, 0.044306587470652775, -0.025052140711712072, -0.048364973237592736, 0.04357424718188166, 0.07329505723450962, -0.015585116770410403, -0.08861317494130555, 0.07579721988781095, 0.12266699836916264, -0.06554445584501527, 0.013075325572434376, -0.0634694916934971, 0.15488997107509195, -0.03197885927633892, 0.011793730067084916, -0.03197885927633892, 0.04601538147778539, -0.008597369848385964, 0.005305652821253272, -0.12352139537272895, 0.0487006292032795, 0.044520186721544355, -0.049738111279038585, 0.05754974102593054, 0.06383566183788265, 0.027798416794603775, 0.006373649075711156, 0.07933686461687137, -0.027447503739567614, 0.050317880674315724, -0.052881071685014644, -0.057915911170316095, -0.036433929366363235, -0.043116534501399705, 0.029049498121254437, -0.011496216824771649, 0.06108938575499095, -0.16196926053321278, -0.0449778994020263, 0.07872658104289543, -0.08568381378622107, 0.03649495772376083, 0.032192458527230496, -0.02535728249870004, -0.05727511341764137, 0.014082293469494668, -0.03463359282313423, 0.09013888387624538, -0.02139043926785647, 0.0357931316136885, -0.08501250185484754, -0.056939457451954605, 0.043208077037496094, -0.12657281324260863, -0.007826886836241349, 0.011831872790458412, -0.015852115834024873, -0.054254209726460494, 0.07665161689137726, -0.08013023326304008, -0.0660937110615936, -0.05233181646843631, 0.10557905829783651, 0.03774603905041149, 0.040278715882411616, 0.025204711605206057, 0.1259625296686327, -0.03628135847286925, -0.040644886026797174, 0.10478568965166779, 0.03960740395103809, -0.06853484535749733, -0.049951710529930166, -0.165386848547478, 0.12095820436203003, 0.06725324985214788, 0.04122465542207431, -0.01820170759383222, -0.06124195664848493, 0.04101105617118274, -5.163571176687E-4, -0.021817637769639624, 0.06462903048405137, 0.05825156713600286, -0.07109803636819625, -0.056725858201063024, 0.11406199997610199, 0.016035200906217656, -0.10356512250371593, 0.06008241785793066, -0.07250168858834091, 0.008955911448096826, 0.014776491034892291, 0.040858485277688754, 0.09026094059104058, -0.13877848472212728, -0.05217924557494232, -0.010237506953446286, -0.039302262164050125, 0.11015618510265601, -0.04235368003392979, 0.045099956116821494, -0.01873570572106116, -0.0040011716818797146, -0.021726095233543234, -0.059411105926557135, -0.05770231191942452, 0.0964858330455951, -0.11088852539142713, -0.024930083996916888, 0.007224231806940114, 0.05339981272289419, -0.04381836061147203, -0.06002138950053307, -0.024762256014073506, 0.03872249276877299, 0.10673859708839079, -0.05187410378795435, 0.023679002670266222, 0.09001682716145021, 0.029247840282796617, 0.1369476340001995, 0.036617014438556014, -0.016050457995567053, 0.05632917387797867, -0.020886955319326328, -0.03680009951074879, 0.024426600048386742, -0.026150651144868753, 0.01992575869031423]] }],
"num_rows": 1
}
'

**Comment by larry88** [comment]
<img width="1229" alt="image" src="https://user-images.githubusercontent.com/11516545/225590434-99dce1f7-bd7b-4d5b-9707-bdf4e29cea63.png">

**Comment by xiaofan-luan** [comment]
we'll be looking into the issue

**Comment by XuanYang-cn** [comment]
Insert data of dim 256 into collection with dim 768 causing this crash, which is unrecoverable for DataNode, cuz data in Pulsar are wrong.

To recover the cluster: drop the collection and create a new collection, and re-insert the right data. The data in pulsar are corrupted.

To prevent this from happening again: recommending using fully-tested sdks for example golang, pymilvus, or java.

Also we might need to add insert data correctness check in proxy @xiaofan-luan

**Comment by xiaofan-luan** [comment]
> Insert data of dim 256 into collection with dim 768 causing this crash, which is unrecoverable for DataNode, cuz data in Pulsar are wrong.
> 
> To recover the cluster: drop the collection and create a new collection, and re-insert the right data. The data in pulsar are corrupted.
> 
> To prevent this from happening again: recommending using fully-tested sdks for example golang, pymilvus, or java.
> 
> Also we might need to add insert data correctness check in proxy @xiaofan-luan

exactly, should be verified on proxy.

**Comment by longjiquan** [comment]
Fixed in #22819 and #22826 .
